File: modules/crewroll-transfer/ios/IdentityKeys/Sources/CrewRollNativeKeys/MediaCrypto/NativePhotoCrypto.swift
import Clibsodium
import CryptoKit
import Foundation

/// File-based v1 framing. Only one 256 KiB plaintext frame is resident at a time.
public enum NativePhotoCrypto {
    private static let frameBytes = 262_144

    public static func randomKey() throws -> Bytes {
        try CrewRollSodium.initialize()
        var key = Bytes(repeating: 0, count: 32)
        randombytes_buf(&key, key.count)
        return key
    }

    public static func mediaKey(contentRoot: Bytes, variant: MediaVariant) throws -> Bytes {
        try requireWidth(contentRoot, 32, "content root")
        var key = Bytes(repeating: 0, count: 32)
        let context = Bytes("CRROLL01".utf8).map { Int8(bitPattern: $0) }
        guard crypto_kdf_derive_from_key(&key, 32, UInt64(variant.rawValue), context, contentRoot) == 0 else {
            throw NativeKeyError.materialLost
        }
        return key
    }

    public static func encryptFile(
        source: URL, destination: URL, contentRoot: Bytes, tripID: String,
        assetID: String, variant: MediaVariant, mime: String, width: UInt32, height: UInt32
    ) throws -> ManifestVariantDescriptor {
        try CrewRollSodium.initialize()
        var key = try mediaKey(contentRoot: contentRoot, variant: variant)
        defer { sodium_memzero(&key, key.count) }
        let aad = try CrewRollAAD.media(tripID: tripID, assetID: assetID, variant: variant)
        var state = crypto_secretstream_xchacha20poly1305_state()
        defer { withUnsafeMutableBytes(of: &state) { sodium_memzero($0.baseAddress, $0.count) } }
        var header = Bytes(repeating: 0, count: 24)
        guard crypto_secretstream_xchacha20poly1305_init_push(&state, &header, key) == 0 else { throw NativeKeyError.materialLost }
        let input = try FileHandle(forReadingFrom: source)
        defer { try? input.close() }
        let size = try input.seekToEnd()
        try input.seek(toOffset: 0)
        let limit: UInt64 = variant == .preview ? 524_288 : 52_428_800
        let frames = max(1, (size + UInt64(frameBytes) - 1) / UInt64(frameBytes))
        guard size <= limit, 24 + size + frames * 21 <= limit else { throw NativeKeyError.invalidCommand }
        try Data().write(to: destination, options: .withoutOverwriting)
        try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: destination.path)
        let output = try FileHandle(forWritingTo: destination)
        var complete = false
        defer { try? output.close(); if !complete { try? FileManager.default.removeItem(at: destination) } }
        var plainHash = SHA256(), cipherHash = SHA256()
        var written: UInt64 = 0
        func write(_ bytes: Bytes) throws {
            let data = Data(bytes)
            try output.write(contentsOf: data)
            cipherHash.update(data: data)
            written += UInt64(data.count)
        }
        try write(header)
        var remaining = size
        repeat {
            let count = Int(min(UInt64(frameBytes), remaining))
            var data = try input.read(upToCount: count) ?? Data()
            guard data.count == count else { throw NativeKeyError.invalidState }
            defer { data.resetBytes(in: 0..<data.count) }
            plainHash.update(data: data)
            remaining -= UInt64(count)
            var message = Bytes(data)
            defer { sodium_memzero(&message, message.count) }
            var encrypted = Bytes(repeating: 0, count: count + 17)
            var encryptedCount: UInt64 = 0
            let tag: UInt8 = remaining == 0 ? 3 : 0
            guard crypto_secretstream_xchacha20poly1305_push(&state, &encrypted, &encryptedCount, message, UInt64(count), aad, UInt64(aad.count), tag) == 0,
                  encryptedCount == UInt64(encrypted.count) else { throw NativeKeyError.materialLost }
            var prefix = ByteWriter()
            prefix.write(UInt32(encrypted.count))
            try write(prefix.bytes)
            try write(encrypted)
        } while remaining > 0
        guard (try input.read(upToCount: 1) ?? Data()).isEmpty else { throw NativeKeyError.invalidState }
        try output.synchronize()
        complete = true
        return ManifestVariantDescriptor(variant: variant, mime: mime, pixelWidth: width, pixelHeight: height,
            plaintextBytes: size, plaintextSHA256: Bytes(plainHash.finalize()), ciphertextBytes: written, ciphertextSHA256: Bytes(cipherHash.finalize()))
    }

    public static func sealManifest(
        contentRoot: Bytes, capturedAt: Date, preview: ManifestVariantDescriptor,
        original: ManifestVariantDescriptor, tripKey: Bytes, tripID: String, assetID: String
    ) throws -> Bytes {
        try requireWidth(contentRoot, 32, "content root")
        var writer = ByteWriter()
        defer { writer.erase() }
        try writer.writeASCII("CRMANP1\0")
        writer.write(UInt32(1))
        writer.write(contentRoot)
        writer.write(UInt8(1))
        let milliseconds = capturedAt.timeIntervalSince1970 * 1000
        guard milliseconds >= 1, milliseconds <= 253_402_300_799_999 else { throw NativeKeyError.invalidCommand }
        write64(UInt64(milliseconds), to: &writer)
        // No filename is necessary for delivery. Keep metadata minimal.
        writer.write([0, 0])
        for descriptor in [preview, original] {
            let mime = Bytes(descriptor.mime.utf8)
            guard mime.count <= 127 else { throw NativeKeyError.invalidCommand }
            writer.write(descriptor.variant.rawValue)
            writer.write(UInt8(mime.count))
            writer.write(mime)
            writer.write(descriptor.pixelWidth)
            writer.write(descriptor.pixelHeight)
            write64(descriptor.plaintextBytes, to: &writer)
            writer.write(descriptor.plaintextSHA256)
            write64(descriptor.ciphertextBytes, to: &writer)
            writer.write(descriptor.ciphertextSHA256)
        }
        var plaintext = writer.bytes
        defer { sodium_memzero(&plaintext, plaintext.count) }
        _ = try ManifestReader.parse(plaintext)
        var key = try ManifestReader.deriveManifestKey(tripKey)
        defer { sodium_memzero(&key, key.count) }
        let aad = try CrewRollAAD.manifest(tripID: tripID, assetID: assetID)
        var nonce = Bytes(repeating: 0, count: 24)
        randombytes_buf(&nonce, nonce.count)
        var ciphertext = Bytes(repeating: 0, count: plaintext.count + 16)
        var count: UInt64 = 0
        guard crypto_aead_xchacha20poly1305_ietf_encrypt(&ciphertext, &count, plaintext, UInt64(plaintext.count), aad, UInt64(aad.count), nil, nonce, key) == 0,
              count == UInt64(ciphertext.count) else { throw NativeKeyError.materialLost }
        return nonce + ciphertext
    }

    public static func decryptFile(
        source: URL, destination: URL, descriptor: ManifestVariantDescriptor,
        contentRoot: Bytes, tripID: String, assetID: String
    ) throws {
        let limit: UInt64 = descriptor.variant == .preview ? 524_288 : 52_428_800
        guard descriptor.ciphertextBytes <= limit else { throw NativeKeyError.invalidCommand }
        var key = try mediaKey(contentRoot: contentRoot, variant: descriptor.variant)
        defer { sodium_memzero(&key, key.count) }
        let aad = try CrewRollAAD.media(tripID: tripID, assetID: assetID, variant: descriptor.variant)
        let input = try FileHandle(forReadingFrom: source)
        defer { try? input.close() }
        guard try input.seekToEnd() == descriptor.ciphertextBytes else { throw NativeKeyError.invalidEnvelope }
        try input.seek(toOffset: 0)
        var cipherHash = SHA256(), plainHash = SHA256()
        func read(_ count: Int) throws -> Bytes {
            let data = try input.read(upToCount: count) ?? Data()
            guard data.count == count else { throw NativeKeyError.invalidEnvelope }
            cipherHash.update(data: data)
            return Bytes(data)
        }
        let header = try read(24)
        var state = crypto_secretstream_xchacha20poly1305_state()
        defer { withUnsafeMutableBytes(of: &state) { sodium_memzero($0.baseAddress, $0.count) } }
        guard crypto_secretstream_xchacha20poly1305_init_pull(&state, header, key) == 0 else { throw NativeKeyError.invalidEnvelope }
        try Data().write(to: destination, options: .withoutOverwriting)
        try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: destination.path)
        let output = try FileHandle(forWritingTo: destination)
        var complete = false
        defer { try? output.close(); if !complete { try? FileManager.default.removeItem(at: destination) } }
        var consumed: UInt64 = 24, plainBytes: UInt64 = 0
        var final = false
        while consumed < descriptor.ciphertextBytes {
            let prefix = try read(4)
            let length = prefix.reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
            guard (17...262_161).contains(length), consumed + 4 + UInt64(length) <= descriptor.ciphertextBytes else { throw NativeKeyError.invalidEnvelope }
            let ciphertext = try read(Int(length))
            consumed += 4 + UInt64(length)
            let last = consumed == descriptor.ciphertextBytes
            guard last ? (length > 17 || consumed == 45) : length == 262_161 else { throw NativeKeyError.invalidEnvelope }
            var message = Bytes(repeating: 0, count: Int(length) - 17)
            defer { sodium_memzero(&message, message.count) }
            var count: UInt64 = 0
            var tag: UInt8 = 0
            guard crypto_secretstream_xchacha20poly1305_pull(&state, &message, &count, &tag, ciphertext, UInt64(length), aad, UInt64(aad.count)) == 0,
                  count == UInt64(message.count), tag == (last ? 3 : 0) else { throw NativeKeyError.invalidEnvelope }
            let data = Data(message)
            plainHash.update(data: data)
            try output.write(contentsOf: data)
            plainBytes += count
            final = last
        }
        guard final, plainBytes == descriptor.plaintextBytes,
              bytesEqual(Bytes(cipherHash.finalize()), descriptor.ciphertextSHA256),
              bytesEqual(Bytes(plainHash.finalize()), descriptor.plaintextSHA256) else { throw NativeKeyError.invalidEnvelope }
        try output.synchronize()
        complete = true
    }

    private static func write64(_ value: UInt64, to writer: inout ByteWriter) {
        writer.write(UInt32(truncatingIfNeeded: value >> 32))
        writer.write(UInt32(truncatingIfNeeded: value))
    }
}

File: modules/crewroll-transfer/android/src/main/java/com/uankit53/crewroll/transfer/media/NativePhotoCrypto.kt
package com.uankit53.crewroll.transfer.media

import com.goterl.lazysodium.Sodium
import com.goterl.lazysodium.interfaces.SecretStream
import java.io.DataInputStream
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.time.Instant

/** Same v1 wire format as iOS; streams bounded frames, never whole originals. */
class NativePhotoCrypto(private val sodium: Sodium) {
    init { check(sodium.sodium_init() >= 0) }
    fun randomKey() = ByteArray(32).also { sodium.randombytes_buf(it, it.size) }
    fun mediaKey(root: ByteArray, variant: MediaVariant): ByteArray {
        requireWidth(root, 32, "content root")
        return ByteArray(32).also { valid(sodium.crypto_kdf_derive_from_key(it, 32, variant.code.toLong(), "CRROLL01".toByteArray(), root) == 0) }
    }
    private fun limit(variant: MediaVariant) = if (variant == MediaVariant.PREVIEW) 524_288L else 52_428_800L
    private fun valid(condition: Boolean) { if (!condition) throw CryptoReadException(CryptoFailure.AUTHENTICATION, "photo integrity verification failed") }
    private fun erase(state: SecretStream.State) { state.k.fill(0); state.nonce.fill(0); state._pad.fill(0); state.write(); state.pointer.clear(state.size().toLong()) }

    fun encrypt(source: File, destination: File, root: ByteArray, tripId: String, assetId: String,
                variant: MediaVariant, mime: String, width: UInt, height: UInt): ManifestVariantDescriptor {
        val size = source.length()
        valid(size <= limit(variant) && 24 + size + maxOf(1, (size + 262_143) / 262_144) * 21 <= limit(variant))
        val key = mediaKey(root, variant)
        val state = SecretStream.State()
        var created = false
        var complete = false
        try {
            val aad = CrewRollAad.media(tripId, assetId, variant)
            val header = ByteArray(24)
            valid(sodium.crypto_secretstream_xchacha20poly1305_init_push(state, header, key) == 0)
            check(destination.createNewFile()) { "destination exists" }; created = true
            val plainHash = MessageDigest.getInstance("SHA-256")
            val cipherHash = MessageDigest.getInstance("SHA-256")
            var written = 0L
            DataInputStream(source.inputStream().buffered()).use { input ->
                FileOutputStream(destination).use { output ->
                    fun write(bytes: ByteArray) { output.write(bytes); cipherHash.update(bytes); written += bytes.size }
                    write(header)
                    var remaining = size
                    do {
                        val message = ByteArray(minOf(262_144L, remaining).toInt())
                        try {
                            input.readFully(message); plainHash.update(message); remaining -= message.size
                            val ciphertext = ByteArray(message.size + 17); val count = LongArray(1)
                            valid(sodium.crypto_secretstream_xchacha20poly1305_push(state, ciphertext, count, message, message.size.toLong(), aad, aad.size.toLong(), (if (remaining == 0L) 3 else 0).toByte()) == 0)
                            valid(count[0] == ciphertext.size.toLong())
                            write(ByteBuffer.allocate(4).putInt(ciphertext.size).array()); write(ciphertext)
                        } finally { message.fill(0) }
                    } while (remaining > 0)
                    valid(input.read() == -1); output.fd.sync()
                }
            }
            complete = true
            return ManifestVariantDescriptor(variant, mime, width, height, size.toULong(), plainHash.digest(), written.toULong(), cipherHash.digest())
        } finally { key.fill(0); erase(state); if (created && !complete) destination.delete() }
    }

    fun decrypt(source: File, destination: File, descriptor: ManifestVariantDescriptor, root: ByteArray, tripId: String, assetId: String) {
        valid(descriptor.ciphertextBytes <= limit(descriptor.variant).toULong() && source.length().toULong() == descriptor.ciphertextBytes)
        val key = mediaKey(root, descriptor.variant); val state = SecretStream.State()
        var created = false; var complete = false
        try {
            val aad = CrewRollAad.media(tripId, assetId, descriptor.variant)
            val cipherHash = MessageDigest.getInstance("SHA-256"); val plainHash = MessageDigest.getInstance("SHA-256")
            DataInputStream(source.inputStream().buffered()).use { input ->
                var consumed = 0L
                fun read(count: Int): ByteArray = ByteArray(count).also { input.readFully(it); consumed += count; cipherHash.update(it) }
                valid(sodium.crypto_secretstream_xchacha20poly1305_init_pull(state, read(24), key) == 0)
                check(destination.createNewFile()) { "destination exists" }; created = true
                var plainBytes = 0L; var final = false
                FileOutputStream(destination).use { output ->
                    while (consumed < source.length()) {
                        val length = ByteBuffer.wrap(read(4)).int
                        valid(length in 17..262_161 && consumed + length <= source.length())
                        val ciphertext = read(length); val last = consumed == source.length()
                        valid(if (last) length > 17 || consumed == 45L else length == 262_161)
                        val message = ByteArray(length - 17); val count = LongArray(1); val tag = ByteArray(1)
                        try {
                            valid(sodium.crypto_secretstream_xchacha20poly1305_pull(state, message, count, tag, ciphertext, length.toLong(), aad, aad.size.toLong()) == 0)
                            valid(count[0] == message.size.toLong() && tag[0].toInt() == if (last) 3 else 0)
                            output.write(message); plainHash.update(message); plainBytes += message.size; final = last
                        } finally { message.fill(0) }
                    }
                    valid(final && input.read() == -1 && plainBytes.toULong() == descriptor.plaintextBytes && bytesEqual(cipherHash.digest(), descriptor.ciphertextSha256) && bytesEqual(plainHash.digest(), descriptor.plaintextSha256))
                    output.fd.sync()
                }
            }
            complete = true
        } finally { key.fill(0); erase(state); if (created && !complete) destination.delete() }
    }

    fun sealManifest(root: ByteArray, captured: Instant, preview: ManifestVariantDescriptor, original: ManifestVariantDescriptor, tripKey: ByteArray, tripId: String, assetId: String): ByteArray {
        requireWidth(root, 32, "content root")
        val buffer = ByteBuffer.allocate(744)
        var plain: ByteArray? = null
        val key = ManifestReader.deriveManifestKey(sodium, tripKey)
        try {
            buffer.put("CRMANP1\u0000".toByteArray()).putInt(1).put(root).put(1).putLong(captured.toEpochMilli()).putShort(0)
            for (d in listOf(preview, original)) {
                val mime = d.mime.toByteArray(Charsets.US_ASCII); valid(mime.size <= 127)
                buffer.put(d.variant.code.toByte()).put(mime.size.toByte()).put(mime).putInt(d.pixelWidth.toInt()).putInt(d.pixelHeight.toInt())
                    .putLong(d.plaintextBytes.toLong()).put(d.plaintextSha256).putLong(d.ciphertextBytes.toLong()).put(d.ciphertextSha256)
            }
            val plaintext = buffer.array().copyOf(buffer.position()); plain = plaintext
            val parsed = ManifestReader.parse(plaintext.copyOf()); parsed.erase()
            val aad = CrewRollAad.manifest(tripId, assetId)
            val nonce = ByteArray(24).also { sodium.randombytes_buf(it, it.size) }
            val ciphertext = ByteArray(plaintext.size + 16); val count = LongArray(1)
            valid(sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(ciphertext, count, plaintext, plaintext.size.toLong(), aad, aad.size.toLong(), null, nonce, key) == 0)
            valid(count[0] == ciphertext.size.toLong())
            return nonce + ciphertext
        } finally { key.fill(0); plain?.fill(0); buffer.array().fill(0) }
    }
    fun openManifest(bytes: ByteArray, tripKey: ByteArray, tripId: String, assetId: String) = ManifestReader.open(sodium, bytes, tripKey, CrewRollAad.manifest(tripId, assetId))
}
